src/iop/atrous.c

```diff
@@ -572,33 +599,28 @@ static void eaw_synthesize(float *const out, const float *const in, const float
 
 #if defined(__SSE2__)
 static void eaw_synthesize_sse2(float *const out, const float *const in, const float *const detail,
-                                const float *thrsf, const float *boostf, const int32_t width,
-                                const int32_t height)
+                                const float *thrsf, const float *boostf, const size_t width,
+                                const size_t height)
 {
-  const __m128 threshold = _mm_set_ps(thrsf[3], thrsf[2], thrsf[1], thrsf[0]);
-  const __m128 boost = _mm_set_ps(boostf[3], boostf[2], boostf[1], boostf[0]);
+  const __m128 threshold = _mm_loadu_ps(thrsf);
+  const __m128 boost = _mm_loadu_ps(boostf);
+  const __m128 zeros = _mm_setzero_ps();
+  const __m128i maski = _mm_set1_epi32(0x80000000u);
+  __m128 *const mask = (__m128 *)&maski;
 
 #ifdef _OPENMP
-#pragma omp parallel for default(none) schedule(static)
+#pragma omp parallel for schedule(static) default(none)
 #endif
-  for(int j = 0; j < height; j++)
+  for(size_t k = 0; k < width * height * 4; k += 4)
   {
-    // TODO: prefetch? _mm_prefetch()
-    const __m128 *pin = (__m128 *)in + (size_t)j * width;
-    __m128 *pdetail = (__m128 *)detail + (size_t)j * width;
-    float *pout = out + (size_t)4 * j * width;
-    for(int i = 0; i < width; i++)
-    {
-      const __m128i maski = _mm_set1_epi32(0x80000000u);
-      const __m128 *mask = (__m128 *)&maski;
-      const __m128 absamt
-          = _mm_max_ps(_mm_setzero_ps(), _mm_sub_ps(_mm_andnot_ps(*mask, *pdetail), threshold));
-      const __m128 amount = _mm_or_ps(_mm_and_ps(*pdetail, *mask), absamt);
-      _mm_stream_ps(pout, _mm_add_ps(*pin, _mm_mul_ps(boost, amount)));
-      pdetail++;
-      pin++;
-      pout += 4;
-    }
+    const __m128 pin = _mm_load_ps(in + k);
+    const __m128 pdetail = _mm_load_ps(detail + k);
+    float *pout = out + k;
+
+    const __m128 absamt
+        = _mm_max_ps(zeros, (_mm_andnot_ps(*mask, pdetail) - threshold));
```

Comment:
Again, I'm not sure about using - instead of _mm_sub_ps

Reply:
you can ask @heckflosse about that

Reply:
Using '-', '+', '*', '/', '-=' .... works fine with gcc and clang, increases readability and also (at least in my experience) gives sometimes faster code (for example a + b * c can give an fma (if supported) while '_mm_add_ps(a, _mm_mul_ps(b,c)' or vice versa in my tests was translated following exactly the intrinsics.

Though I have to admit, that the last time I looked at the assembly is a while ago.

Anyway, in RT I always use the more readable notation and try to avoid intrinsics and never experienced problems, neither with clang, nor with gcc.

One exception is the following, which is supported by gcc, but not by clang:

```
__m128 a;
float b;

a = a * b;
```

Reply:
@heckflosse : thanks for the input. indeed the float * _m128 is not supported by clang, we had just fixed this in filmic.